### PR TITLE
[Imporve][Fake-Connector-V2]support user-defined-schmea and random data for fake-table 

### PR DESCRIPTION
**Pull request by laglangyue**
## Purpose of this pull request

the connector-v2-fake did't not support enough data type, and not quickly  to modify，more time we need the fake to test the dataType Mapper，
so I do these works：
(1) optimization connector-fake-v2 so that we can easy to add column name,type and data
(2) change the spark example for connector-fake-v2，and fix some dead Http link in this example conf file
(3) I find the method don't contain the condition SqlType=NULL： org.apache.seatunnel.translation.serialization.RowConverter#validate，I don't sure it OK。and the SqlType of VOID_TYPE  is NULL
(4) the error message is bad in org.apache.seatunnel.translation.serialization.RowConverter(line 60), i fix it

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
+package org.apache.seatunnel.connectors.seatunnel.fake.source;
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java]: The license is missing.

Comment (laglangyue) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java]: done

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.commons.lang3.RandomUtils;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+
+public class FakeData {
+
+    public static final String[] COLUMN_NAME = new String[]{
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java]: Can you support user-defined schema fields?  Please reference #2392

Comment (laglangyue) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java]: ok，I try it，but may be in somtime, and no later than the weekend. I don't have much time on weekdays.

**Comment by laglangyue** [comment]
I find `SeatunnelSchema `that it not implement `Serializable`, it Example-task will error.
I add it.

**Comment by laglangyue** [comment]
the PR has completed. I add Junit5-test and test random data correctly.
For some complex types, example-v2-spark may occur some exceptions, but they are not the reason for Fake-V2, and during the spark type conversion process. I will test and try to repair them in another PR, because there are too many in this PR

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java (hunk 2 above)
Comment (laglangyue) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeData.java]: > Can you support user-defined schema fields? Please reference #2392
done

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf [hunk 3]
```diff
@@ -45,20 +62,20 @@ source {
...
   # you can also use other transform plugins, such as sql
   sql {
-    sql = "select name,age from fake"
+    sql = "select c_string,c_boolean,c_tinyint,c_smallint,c_int,c_bigint,c_float,c_double,c_null,c_bytes from fake"
     result_table_name = "sql"
   }
 
   # If you would like to get more information about how to configure seatunnel and see full list of transform plugins,
-  # please go to https://seatunnel.apache.org/docs/spark/configuration/transform-plugins/Split
+  # please go to https://seatunnel.apache.org/docs/2.1.3/category/transform
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: Use https://seatunnel.apache.org/docs/category/transform ?

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf [hunk 4]
```diff
@@ -45,20 +62,20 @@ source {
   # }
 
   # If you would like to get more information about how to configure seatunnel and see full list of input plugins,
-  # please go to https://seatunnel.apache.org/docs/spark/configuration/source-plugins/Fake
+  # please go to https://seatunnel.apache.org/docs/2.1.3/category/source
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: Use https://seatunnel.apache.org/docs/category/source-v2 ?

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf [hunk 5]
```diff
@@ -72,5 +89,5 @@ sink {
   # }
 
   # If you would like to get more information about how to configure seatunnel and see full list of output plugins,
-  # please go to https://seatunnel.apache.org/docs/spark/configuration/sink-plugins/Console
+  # please go to https://seatunnel.apache.org/docs/2.1.3/category/sink
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: Use https://seatunnel.apache.org/docs/category/sink-v2 ?

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf [hunk 6]
```diff
@@ -33,8 +33,25 @@ env {
...
+      c_decimal = "decimal(30, 8)"
+      c_null = "null"
+      c_bytes = bytes
+      c_date = date
+      c_time = time
+      c_timestamp = timestamp
+    }
     result_table_name = "fake"
-    field_name = "name,age,timestamp"
   }
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: Should you change all testcase?  

https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-examples/seatunnel-flink-connector-v2-example/src/main/resources/examples
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-spark-connector-v2-e2e/src/test/resources

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            Object value = randomColumnValue(valueType);
+            HashMap<Object, Object> objectObjectHashMap = new HashMap<>();
+            objectObjectHashMap.put(key, value);
+            return objectObjectHashMap;
+        } else if (fieldType instanceof PrimitiveByteArrayType) {
+            return RandomUtils.nextBytes(10);
+        } else if (VOID_TYPE.equals(fieldType) || fieldType == null) {
+            return Void.TYPE;
+        } else {
+            throw new IllegalStateException("Unexpected value: " + fieldType);
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java]: Change to UnsupportedOperationException？

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        } else if (LocalTimeType.LOCAL_DATE_TYPE.equals(fieldType)) {
+            return randomLocalDateTime().toLocalDate();
+        } else if (LocalTimeType.LOCAL_TIME_TYPE.equals(fieldType)) {
+            return randomLocalDateTime().toLocalTime();
+        } else if (LocalTimeType.LOCAL_DATE_TIME_TYPE.equals(fieldType)) {
+            return randomLocalDateTime();
+        } else if (fieldType instanceof DecimalType) {
+            DecimalType decimalType = (DecimalType) fieldType;
+            return new BigDecimal(RandomStringUtils.randomNumeric(decimalType.getPrecision() - decimalType.getScale()) + "." +
+                RandomStringUtils.randomNumeric(decimalType.getPrecision() - decimalType.getScale()));
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java]: Should used `RandomStringUtils.randomNumeric(decimalType.getScale())` ?

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        }
+        return new SeaTunnelRow(randomRow.toArray());
+    }
+
+    @SuppressWarnings("magicnumber")
+    private Object randomColumnValue(SeaTunnelDataType<?> fieldType) {
+        if (BOOLEAN_TYPE.equals(fieldType)) {
+            return RandomUtils.nextInt(0, 2) == 1;
+        } else if (BYTE_TYPE.equals(fieldType)) {
+            return (byte) RandomUtils.nextInt(0, Byte.MAX_VALUE);
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java]: Should used `RandomUtils.nextInt(Byte.MIN_VALUE, Byte.MAX_VALUE)` or `RandomUtils.nextInt(0, 255)` ?

**Review by hailin0 — COMMENTED** [review]
Can you add the connectors-v2/FakeSource docs?

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java (hunk 9 above)
Comment (laglangyue) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java]: done

File: seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java (hunk 7 above)
Comment (laglangyue) [reply seatunnel-connectors-v2/connector-fake/src/main/java/org/apache/seatunnel/connectors/seatunnel/fake/source/FakeRandomData.java]: > Change to UnsupportedOperationException？
done

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf (hunk 3 above)
Comment (laglangyue) [reply seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: done

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf (hunk 5 above)
Comment (laglangyue) [reply seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: done

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf (hunk 6 above)
Comment (laglangyue) [reply seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: yes, I do, but I have not test it, I will test as soon as possible。

Comment (laglangyue) [reply seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples/spark.batch.conf]: > Should you change all testcase?
> 
> https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-examples/seatunnel-flink-connector-v2-example/src/main/resources/examples https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/resources/examples https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-spark-connector-v2-e2e/src/test/resources

Thx，I have changed all of them. Is there has  auto-test by CI，the test is so much.

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf [hunk 10]
```diff
@@ -25,64 +25,66 @@ env {
...
+      c_bigint = bigint
+      c_float = float
+      c_double = double
+      c_decimal = "decimal(30, 8)"
+      c_null = "null"
+      c_bytes = bytes
+      c_date = date
+      c_time = time
+      c_timestamp = timestamp
     }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf]: For all testcase & examples just change the FakeSource fields code style:

from:
FakeSource {
       result_table_name = "fake"
       field_name = "name,age"
}

to:
FakeSource {
       result_table_name = "fake"
       fields {
         name = string
         age = int
       }
}

Comment (laglangyue) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf]: thx，good idea.

**Comment by hailin0** [comment]
> Can you add the connectors-v2/FakeSource docs?

here https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2/source

**Comment by laglangyue** [comment]
> Can you add the connectors-v2/FakeSource docs?

yes ,  I update the config to keep the schema parameters consistent such as LocalFile and HDFS source.
Thank you for your guidance. This is my first contribution in Github.

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf [hunk 11]
```diff
@@ -30,11 +30,16 @@ source {
   # This is a example source plugin **only for test and demonstrate the feature source plugin**
     FakeSource {
       result_table_name = "fake"
-      field_name = "name"
+      schema = {
+        fields {
+          name = "string"
+          age = "int"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf]: remove  `age` field ?

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf [hunk 12]
```diff
@@ -84,5 +88,5 @@ sink {
            ]
    }
   # If you would like to get more information about how to configure seatunnel and see full list of sink plugins,
-  # please go to https://seatunnel.apache.org/docs/flink/configuration/sink-plugins/Console
+  # please go to https://seatunnel.apache.org/docs/connector-v2/sink/Assert
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf:91]: update Assert rules

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf (hunk 11 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/jdbc/fakesource_to_jdbc.conf]: done

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf (hunk 12 above)
Comment (laglangyue) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/src/test/resources/assertion/fakesource_to_assert.conf:91]: done

**Comment by hailin0** [comment]
LGTM
@CalvinKirs

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**